### DISCUSSION TITLE
Does GLSP have an automated test framework?

### DISCUSSION THREAD

**Discussion by gee4vee** [eclipse-glsp/glsp#737] [General] [answered]
Is there a framework for writing automated tests against a rendered diagram? If so, does it support the various clients (VS Code, Theia, browser)?

**Comment by planger** [comment, marked as answer]
No, there is no dedicated framework (yet).

For unit and integration testing (e.g. handlers, factories, etc.), we typically use plain JUnit or mocha, depending on the server in our projects.
For end-to-end tests, we used Playwright quite successfully in some projects. In those projects, we usually also write a page object model library to assist writing more concise tests, similar to [Theia Playwright](https://github.com/eclipse-theia/theia/tree/master/examples/playwright). We didn't extract a generic framework out of those page object models yet though.
Also, we only used Playwright for browser testing so far, but not for Electron. Theoretically there is experimental support though.

However, there might be something along those lines coming up in the next months. If you are interested in contributing to that endeavor please let us know!

**Reply by Rajsmit** [reply]
Hi @planger, Could you please provide some example or link where Playwright is used to test GLSP diagrams?

**Reply by planger** [reply]
Hi, at the moment I'm not aware of any public usages of Playwright for testing GLSP diagrams. We are using Playwright as described above for GLSP diagrams in proprietary products though, which I can't share, unfortunately.

**Reply by Rajsmit** [reply]
@planger  Thanks for information.
